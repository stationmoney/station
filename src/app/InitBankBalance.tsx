import { PropsWithChildren } from "react"
import {
  useInitialBankBalance,
  useInitialTokenBalance,
} from "data/queries/bank"
import { BankBalanceProvider } from "data/queries/bank"
import NetworkLoading from "./NetworkLoading"

const InitBankBalance = ({ children }: PropsWithChildren<{}>) => {
  const { data: bankBalance } = useInitialBankBalance()
  const { data: tokenBalance } = useInitialTokenBalance()
  // If the balance doesn't exist, nothing is worth rendering.
  if (!bankBalance) return <NetworkLoading title="Fetching balances..." />
  return (
    <BankBalanceProvider value={[...bankBalance, ...(tokenBalance ?? [])]}>
      {children}
    </BankBalanceProvider>
  )
}

export default InitBankBalance
