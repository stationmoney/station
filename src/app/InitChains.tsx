import axios from "axios"
import { ASSETS } from "config/constants"
import { WhitelistProvider } from "data/queries/chains"
import { PropsWithChildren, useEffect, useState } from "react"

type Whitelist = Record<
  string,
  {
    token: string
    symbol: string
    name: string
    icon: string
    chains: string[]
    decimals: number
  }
>

const InitChains = ({ children }: PropsWithChildren<{}>) => {
  const [data, setData] = useState<Whitelist>()
  useEffect(() => {
    ;(async () => {
      const { data: whitelist } = await axios.get("/station/coins.json", {
        baseURL: ASSETS,
      })
      setData(whitelist)
    })()
  }, [])

  if (!data) return null

  return <WhitelistProvider value={data}>{children}</WhitelistProvider>
}

export default InitChains
