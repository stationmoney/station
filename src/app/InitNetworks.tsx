import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
