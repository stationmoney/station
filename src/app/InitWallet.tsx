import { PropsWithChildren, useEffect, useMemo } from "react"
import { QueryClient, QueryClientProvider } from "react-query"
import { WalletStatus } from "@terra-money/wallet-types"
import { useWallet } from "@terra-money/use-wallet"
import { useNetworkName } from "data/wallet"
import { isWallet, useAuth } from "auth"
import Online from "./containers/Online"
import NetworkLoading from "./NetworkLoading"

const InitWallet = ({ children }: PropsWithChildren<{}>) => {
  useOnNetworkChange()
  const { status } = useWallet()
  const queryClient = useQueryClient()
  const networkName = useNetworkName()

  return status === WalletStatus.INITIALIZING ? (
    <NetworkLoading title="Initializing your wallet..." />
  ) : (
    <QueryClientProvider client={queryClient} key={networkName}>
      {children}
      <Online />
    </QueryClientProvider>
  )
}

export default InitWallet

/* hooks */
const useOnNetworkChange = () => {
  const { wallet, disconnect } = useAuth()
  const isPreconfiguredWallet = isWallet.preconfigured(wallet)
  const shouldDisconnect = isPreconfiguredWallet

  useEffect(() => {
    if (shouldDisconnect) disconnect()
  }, [disconnect, shouldDisconnect])
}

const useQueryClient = () => {
  const name = useNetworkName()

  return useMemo(() => {
    if (!name) throw new Error()
    return new QueryClient()
  }, [name])
}
