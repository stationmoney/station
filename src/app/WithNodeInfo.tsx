import { PropsWithChildren } from "react"
import { useNodeInfo } from "data/queries/tendermint"
import Overlay from "./components/Overlay"
import NetworkError from "./NetworkError"
import { useNetworkName } from "data/wallet"

const WithNodeInfo = ({ children }: PropsWithChildren<{}>) => {
  const network = useNetworkName()
  // FIXME: remove or make sure chainID is available
  const { isLoading, isError } = useNodeInfo(
    network === "mainnet" ? "phoenix-1" : "pisco-1"
  )

  if (isError) {
    return (
      <Overlay>
        <NetworkError />
      </Overlay>
    )
  }

  if (isLoading) return null
  return <>{children}</>
}

export default WithNodeInfo
