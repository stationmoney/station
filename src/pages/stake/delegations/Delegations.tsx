import { useTranslation } from "react-i18next"
import { AccAddress, Coin, Delegation } from "@terra-money/feather.js"
import { getMaxHeightStyle } from "utils/style"
import { combineState } from "data/query"
import { useMemoizedPrices } from "data/queries/coingecko"
import { useInterchainDelegations } from "data/queries/staking"
import { ValidatorLink } from "components/general"
import { ModalButton } from "components/feedback"
import { Table } from "components/layout"
import { Read } from "components/token"
import StakedCard from "../components/StakedCard"
import { useNativeDenoms } from "data/token"
import styles from "../CardModal.module.scss"

const Delegations = () => {
  const { t } = useTranslation()
  const readNativeDenom = useNativeDenoms()
  const { data: prices, ...pricesState } = useMemoizedPrices()

  const interchainDelegations = useInterchainDelegations()

  const delegations: Delegation[] = interchainDelegations.reduce(
    (acc, { data }) => (data ? [...data?.delegation, ...acc] : acc),
    [] as Delegation[]
  )
  const state = combineState(pricesState, ...interchainDelegations)

  /* render */
  const title = t("Delegations")

  const render = () => {
    if (!delegations || !prices) return null

    const total = delegations.reduce((acc, { balance }) => {
      const { token, decimals } = readNativeDenom(balance.denom)
      return (
        acc +
        (balance.amount.toNumber() * (prices[token]?.price || 0)) /
          10 ** decimals
      )
    }, 0)

    return (
      <ModalButton
        title={title}
        renderButton={(open) => (
          <StakedCard
            {...state}
            title={
              <div className={styles.header_wrapper}>
                {title}
                {total !== -1 && (
                  <span className={styles.view_more}>View More</span>
                )}
              </div>
            }
            value={total?.toString() || "0"}
            onClick={open}
            cardName={"delegations"}
          />
        )}
      >
        <Table
          dataSource={delegations}
          sorter={({ balance: { amount: a } }, { balance: { amount: b } }) =>
            b.minus(a).toNumber()
          }
          columns={[
            {
              title: t("Validator"),
              dataIndex: "validator_address",
              render: (address: AccAddress) => (
                <ValidatorLink address={address} internal />
              ),
            },
            {
              title: t("Delegated"),
              dataIndex: "balance",
              render: (balance: Coin) => <Read {...balance.toData()} />,
              align: "right",
            },
          ]}
          style={getMaxHeightStyle(320)}
        />
      </ModalButton>
    )
  }

  return render()
}

export default Delegations
