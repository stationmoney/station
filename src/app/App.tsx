import { getErrorMessage } from "utils/error"
import Layout, { MainContainer, Page } from "components/layout"
import { Banner, Content, Header, Actions, Sidebar } from "components/layout"
import { ErrorBoundary, Wrong } from "components/feedback"

/* routes */
import { useNav } from "./routes"

/* banner */
import UpdateExtension from "./sections/UpdateExtension"

/* sidebar */
import Nav from "./sections/Nav"
import Aside from "./sections/Aside"

/* header */
import NetworkHeader from "./sections/NetworkHeader"
import Refresh from "./sections/Refresh"
import Preferences from "./sections/Preferences"
import ConnectWallet from "./sections/ConnectWallet"

/* extra */
import LatestTx from "./sections/LatestTx"
import ValidatorButton from "./sections/ValidatorButton"
import DevTools from "./sections/DevTools"
import { useShowWelcomeModal } from "utils/localStorage"

/* init */
import InitBankBalance from "./InitBankBalance"
import Wallet from "pages/wallet/Wallet"
import WelcomeModal from "./sections/WelcomeModal"
import NavButton from "./sections/NavButton"
import NetworkStatus from "components/display/NetworkStatus"

const App = () => {
  const { element: routes } = useNav()
  const showModal = useShowWelcomeModal()

  return (
    <Layout>
      <Banner>
        <UpdateExtension />
      </Banner>

      <Sidebar>
        <Nav />
        <Aside />
      </Sidebar>

      <Header>
        <NetworkHeader />

        <Actions>
          <DevTools />
          <section>
            <Refresh />
            <Preferences />
            <NetworkStatus />
          </section>
          <ValidatorButton />
          <ConnectWallet />
          <NavButton />
        </Actions>
        <LatestTx />
      </Header>

      <Content>
        <ErrorBoundary fallback={fallback}>
          <InitBankBalance>
            <MainContainer>
              {routes}
              <Wallet />
              {showModal && <WelcomeModal />}
            </MainContainer>
          </InitBankBalance>
        </ErrorBoundary>
      </Content>
    </Layout>
  )
}

export default App

/* error */
export const fallback = (error: Error) => (
  <Page>
    <Wrong>{getErrorMessage(error)}</Wrong>
  </Page>
)
