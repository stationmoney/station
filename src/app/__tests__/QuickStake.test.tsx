import React from "react"
import { useQuery } from "react-query"
import { render, screen } from "@testing-library/react"
import user from "@testing-library/user-event"
import QuickStake from "pages/stake/QuickStake"
import { NetworksProvider } from "app/InitNetworks"
import { MemoryRouter } from "react-router-dom"
import { RecoilRoot } from "recoil"
import { mockNetworks } from "./__mocks__/Networks.mock"
import { mockValidators, mockPriorityVals } from "./__mocks__/Validators.mock"
import { mockDelegations } from "./__mocks__/Delegations.mock"
import { mockBalances } from "./__mocks__/Balances.mock"
import { mockBankBalance } from "./__mocks__/BankBalance.mock"
import { mockWhitelist } from "./__mocks__/Whitelist.mock"
import { mockExchangeRates } from "./__mocks__/ExchangeRates.mock"
import { mockStakingParams } from "./__mocks__/StakingParams.mock"
import { mockTaxRate } from "./__mocks__/TaxRate.mock"
import { mockTaxCap } from "./__mocks__/TaxCap.mock"

function renderComponent() {
  type TokenFilter = <T>(network: Record<string, T>) => Record<string, T>
  const networks = {} as jest.Mocked<InterchainNetworks>
  const mockedTokenFilter = jest.fn() as jest.Mocked<TokenFilter>

  return render(
    <NetworksProvider
      value={{
        networksLoading: false,
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <MemoryRouter>
        <RecoilRoot>
          <QuickStake />
        </RecoilRoot>
      </MemoryRouter>
    </NetworksProvider>
  )
}

jest.mock("../../data/wallet", () => {
  const mockUseAddress = () => {
    return "terra111111111111111111111111111111111111111"
  }

  const mockUseNetwork = () => {
    return mockNetworks
  }

  const mockUseNetworkName = () => {
    return "testnet"
  }

  const mockUseChainID = () => {
    return "pisco-1"
  }

  return {
    useAddress: mockUseAddress,
    useNetwork: mockUseNetwork,
    useChainID: mockUseChainID,
    useNetworkName: mockUseNetworkName,
  }
})

jest.mock("../../data/queries/staking", () => {
  const mockUseStakingParams = () => mockStakingParams
  const mockUseDelegations = () => mockDelegations
  const mockUseValidators = () => mockValidators
  const mockGetPriorityVals = () => mockPriorityVals
  const mockGetChainUnbondTime = () => 5
  const mockCalcDelegationsTotal = () => 12899999

  return {
    useStakingParams: mockUseStakingParams,
    useDelegations: mockUseDelegations,
    useValidators: mockUseValidators,
    getPriorityVals: mockGetPriorityVals,
    getChainUnbondTime: mockGetChainUnbondTime,
    calcDelegationsTotal: mockCalcDelegationsTotal,
  }
})

jest.mock("../../data/queries/bank", () => {
  const mockUseIsWalletEmpty = () => {
    return false
  }

  const mockUseBankBalance = () => {
    return mockBankBalance
  }

  const mockUseBalances = () => {
    return mockBalances
  }

  return {
    useIsWalletEmpty: mockUseIsWalletEmpty,
    useBankBalance: mockUseBankBalance,
    useBalances: mockUseBalances,
  }
})

jest.mock("../../data/queries/chains", () => {
  const mockUseWhitelist = () => {
    return mockWhitelist
  }

  return {
    useWhitelist: mockUseWhitelist,
  }
})

jest.mock("../../data/queries/coingecko", () => {
  const mockUseExchangeRates = () => {
    return mockExchangeRates
  }

  return {
    useExchangeRates: mockUseExchangeRates,
  }
})

jest.mock("../../data/queries/treasury", () => {
  const mockUseTaxRate = () => {
    return mockTaxRate
  }

  const mockUseTaxCap = () => {
    return mockTaxCap
  }

  return {
    useTaxRate: mockUseTaxRate,
    useTaxCap: mockUseTaxCap,
  }
})

jest.mock("@terra-money/terra-utils", () => {
  const actual = jest.requireActual("@terra-money/terra-utils")
  const mockTruncate = (value: string) => value

  return {
    ...actual,
    truncate: mockTruncate,
  }
})

jest.mock("react-query", () => ({
  useQuery: jest.fn().mockReturnValue({
    data: Object,
    isLoading: Boolean,
    error: Object,
  }),
}))

describe("QuickStake", async () => {
  beforeEach(() => {
    useQuery.mockReturnValue({
      data: 420829,
      isLoading: false,
      error: null,
    })
  })

  it("matches original component", () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })

  it("matches user augmented component", async () => {
    const { asFragment } = renderComponent()

    const networksButton = screen.getByRole("button", {
      name: "+ 5",
    })
    await user.click(networksButton)

    expect(asFragment()).toMatchSnapshot()
  })
})
