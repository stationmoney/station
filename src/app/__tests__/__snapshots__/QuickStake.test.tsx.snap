// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickStake matches original component 1`] = `
<DocumentFragment>
  <article
    class="page"
  >
    <div
      class="container grid"
    >
      <section
        class="main"
      >
        <div
          class="buttonfilter"
        >
          <div
            class="header"
          />
          <div
            class="actions"
          >
            <button
              class="active"
            >
              Delegate
            </button>
            <button>
              Undelegate
            </button>
          </div>
          <div
            class="content"
          >
            <article
              class="card muted default"
            >
              <section
                class="main"
              >
                <article
                  class="page small invisible"
                >
                  <div
                    class="container grid"
                  >
                    <section
                      class="main"
                    >
                      <div
                        class="chainfilter__out"
                      >
                        <div
                          class="filter header"
                        >
                          <div
                            class="pills"
                          >
                            <button
                              aria-expanded="false"
                              class="button"
                            >
                              + 5
                            </button>
                          </div>
                        </div>
                        <div
                          class="content"
                        >
                          <article
                            class="page invisible"
                          >
                            <div
                              class="container grid"
                            >
                              <section
                                class="main"
                              >
                                <form
                                  class="form undefined"
                                >
                                  <div
                                    class="component warning"
                                  >
                                    <div
                                      class="flex"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                        data-testid="InfoIcon"
                                        focusable="false"
                                        style="font-size: 18px;"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                                        />
                                      </svg>
                                    </div>
                                    <span>
                                      Leave enough coins to pay fee for subsequent transactions
                                    </span>
                                  </div>
                                  <div
                                    class="grid"
                                    style="gap: 4px;"
                                  >
                                    <header
                                      class="header"
                                    >
                                      <label
                                        class="label"
                                      >
                                        Amount
                                      </label>
                                      <aside
                                        class="extra"
                                      >
                                        <button
                                          class="muted"
                                          type="button"
                                        >
                                          <div
                                            class="flex start"
                                            style="gap: 4px;"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit icon css-1vooibu-MuiSvgIcon-root"
                                              data-testid="AccountBalanceWalletIcon"
                                              focusable="false"
                                              viewBox="0 0 24 24"
                                            >
                                              <path
                                                d="M21 18v1c0 1.1-.9 2-2 2H5c-1.11 0-2-.9-2-2V5c0-1.1.89-2 2-2h14c1.1 0 2 .9 2 2v1h-9c-1.11 0-2 .9-2 2v8c0 1.1.89 2 2 2h9zm-9-2h10V8H12v8zm4-2.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
                                              />
                                            </svg>
                                            <span
                                              class="component"
                                            >
                                              1,002
                                              <span
                                                class="small"
                                              >
                                                .284039
                                              </span>
                                              <span
                                                class="small"
                                              >
                                                 LUNA
                                              </span>
                                            </span>
                                          </div>
                                        </button>
                                      </aside>
                                    </header>
                                    <div
                                      class="wrapper"
                                    >
                                      <input
                                        autocomplete="off"
                                        class="input before"
                                        inputmode="decimal"
                                        name="input"
                                        placeholder="0.000000"
                                      />
                                      <div
                                        class="flex symbol after"
                                      >
                                        LUNA
                                      </div>
                                    </div>
                                  </div>
                                  <div
                                    class="component"
                                  >
                                    <dl>
                                      <dt
                                        class="gas"
                                      >
                                        Fee
                                      </dt>
                                      <dd>
                                        <span
                                          class="component"
                                        >
                                          0
                                          <span
                                            class="small"
                                          >
                                            .006313
                                          </span>
                                          <span
                                            class="small"
                                          >
                                             LUNA
                                          </span>
                                        </span>
                                      </dd>
                                      <dt>
                                        Balance
                                      </dt>
                                      <dd>
                                        <span
                                          class="component"
                                        >
                                          1,002
                                          <span
                                            class="small"
                                          >
                                            .290352
                                          </span>
                                          <span
                                            class="small"
                                          >
                                             LUNA
                                          </span>
                                        </span>
                                      </dd>
                                      <dt>
                                        Balance after tx
                                      </dt>
                                      <dd>
                                        <span
                                          class="component"
                                        >
                                          1,002
                                          <span
                                            class="small"
                                          >
                                            .284039
                                          </span>
                                          <span
                                            class="small"
                                          >
                                             LUNA
                                          </span>
                                        </span>
                                      </dd>
                                    </dl>
                                  </div>
                                  <button
                                    aria-expanded="false"
                                    class="button small outline button"
                                    type="button"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                                      data-testid="AccountBalanceWalletIcon"
                                      focusable="false"
                                      style="font-size: 16px;"
                                      viewBox="0 0 24 24"
                                    >
                                      <path
                                        d="M21 18v1c0 1.1-.9 2-2 2H5c-1.11 0-2-.9-2-2V5c0-1.1.89-2 2-2h14c1.1 0 2 .9 2 2v1h-9c-1.11 0-2 .9-2 2v8c0 1.1.89 2 2 2h9zm-9-2h10V8H12v8zm4-2.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
                                      />
                                    </svg>
                                    <span
                                      class="button__text"
                                    >
                                      420829
                                    </span>
                                  </button>
                                </form>
                              </section>
                            </div>
                          </article>
                        </div>
                      </div>
                    </section>
                  </div>
                </article>
              </section>
            </article>
          </div>
        </div>
      </section>
    </div>
  </article>
</DocumentFragment>
`;

exports[`QuickStake matches user augmented component 1`] = `
<DocumentFragment>
  <article
    class="page"
  >
    <div
      class="container grid"
    >
      <section
        class="main"
      >
        <div
          class="buttonfilter"
        >
          <div
            class="header"
          />
          <div
            class="actions"
          >
            <button
              class="active"
            >
              Delegate
            </button>
            <button>
              Undelegate
            </button>
          </div>
          <div
            class="content"
          >
            <article
              class="card muted default"
            >
              <section
                class="main"
              >
                <article
                  class="page small invisible"
                >
                  <div
                    class="container grid"
                  >
                    <section
                      class="main"
                    >
                      <div
                        class="chainfilter__out"
                      >
                        <div
                          class="filter header"
                        >
                          <div
                            class="pills"
                          >
                            <button
                              aria-expanded="true"
                              class="button"
                            >
                              + 5
                            </button>
                            <div
                              data-tippy-root=""
                              id="tippy-3"
                              style="z-index: 9999; visibility: visible; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 10px);"
                            >
                              <div
                                class="tippy-box popover"
                                data-escaped=""
                                data-placement="bottom"
                                data-reference-hidden=""
                                data-state="hidden"
                                data-theme="none"
                                role="tooltip"
                                style="max-width: 200px; transition-duration: 0ms;"
                                tabindex="-1"
                              >
                                <div
                                  class="tippy-content"
                                  data-state="hidden"
                                  style="transition-duration: 0ms;"
                                >
                                  <div>
                                    <div
                                      class="options"
                                    >
                                      <button
                                        class="button active"
                                      >
                                        <img
                                          alt="Terra"
                                          src="https://station-assets.terra.money/img/chains/Terra.svg"
                                        />
                                        Terra
                                      </button>
                                      <button
                                        class="button"
                                      >
                                        <img
                                          alt="Atreides"
                                          src="https://station-assets.terra.money/img/chains/Atreides.png"
                                        />
                                        Atreides
                                      </button>
                                      <button
                                        class="button"
                                      >
                                        <img
                                          alt="Corrino"
                                          src="https://station-assets.terra.money/img/chains/Corrino.png"
                                        />
                                        Corrino
                                      </button>
                                      <button
                                        class="button"
                                      >
                                        <img
                                          alt="Harkonnen"
                                          src="https://station-assets.terra.money/img/chains/Harkonnen.png"
                                        />
                                        Harkonnen
                                      </button>
                                      <button
                                        class="button"
                                      >
                                        <img
                                          alt="Ordos"
                                          src="https://station-assets.terra.money/img/chains/Ordos.png"
                                        />
                                        Ordos
                                      </button>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="content"
                        >
                          <article
                            class="page invisible"
                          >
                            <div
                              class="container grid"
                            >
                              <section
                                class="main"
                              >
                                <form
                                  class="form undefined"
                                >
                                  <div
                                    class="component warning"
                                  >
                                    <div
                                      class="flex"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                        data-testid="InfoIcon"
                                        focusable="false"
                                        style="font-size: 18px;"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                                        />
                                      </svg>
                                    </div>
                                    <span>
                                      Leave enough coins to pay fee for subsequent transactions
                                    </span>
                                  </div>
                                  <div
                                    class="grid"
                                    style="gap: 4px;"
                                  >
                                    <header
                                      class="header"
                                    >
                                      <label
                                        class="label"
                                      >
                                        Amount
                                      </label>
                                      <aside
                                        class="extra"
                                      >
                                        <button
                                          class="muted"
                                          type="button"
                                        >
                                          <div
                                            class="flex start"
                                            style="gap: 4px;"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit icon css-1vooibu-MuiSvgIcon-root"
                                              data-testid="AccountBalanceWalletIcon"
                                              focusable="false"
                                              viewBox="0 0 24 24"
                                            >
                                              <path
                                                d="M21 18v1c0 1.1-.9 2-2 2H5c-1.11 0-2-.9-2-2V5c0-1.1.89-2 2-2h14c1.1 0 2 .9 2 2v1h-9c-1.11 0-2 .9-2 2v8c0 1.1.89 2 2 2h9zm-9-2h10V8H12v8zm4-2.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
                                              />
                                            </svg>
                                            <span
                                              class="component"
                                            >
                                              1,002
                                              <span
                                                class="small"
                                              >
                                                .284039
                                              </span>
                                              <span
                                                class="small"
                                              >
                                                 LUNA
                                              </span>
                                            </span>
                                          </div>
                                        </button>
                                      </aside>
                                    </header>
                                    <div
                                      class="wrapper"
                                    >
                                      <input
                                        autocomplete="off"
                                        class="input before"
                                        inputmode="decimal"
                                        name="input"
                                        placeholder="0.000000"
                                      />
                                      <div
                                        class="flex symbol after"
                                      >
                                        LUNA
                                      </div>
                                    </div>
                                  </div>
                                  <div
                                    class="component"
                                  >
                                    <dl>
                                      <dt
                                        class="gas"
                                      >
                                        Fee
                                      </dt>
                                      <dd>
                                        <span
                                          class="component"
                                        >
                                          0
                                          <span
                                            class="small"
                                          >
                                            .006313
                                          </span>
                                          <span
                                            class="small"
                                          >
                                             LUNA
                                          </span>
                                        </span>
                                      </dd>
                                      <dt>
                                        Balance
                                      </dt>
                                      <dd>
                                        <span
                                          class="component"
                                        >
                                          1,002
                                          <span
                                            class="small"
                                          >
                                            .290352
                                          </span>
                                          <span
                                            class="small"
                                          >
                                             LUNA
                                          </span>
                                        </span>
                                      </dd>
                                      <dt>
                                        Balance after tx
                                      </dt>
                                      <dd>
                                        <span
                                          class="component"
                                        >
                                          1,002
                                          <span
                                            class="small"
                                          >
                                            .284039
                                          </span>
                                          <span
                                            class="small"
                                          >
                                             LUNA
                                          </span>
                                        </span>
                                      </dd>
                                    </dl>
                                  </div>
                                  <button
                                    aria-expanded="false"
                                    class="button small outline button"
                                    type="button"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                                      data-testid="AccountBalanceWalletIcon"
                                      focusable="false"
                                      style="font-size: 16px;"
                                      viewBox="0 0 24 24"
                                    >
                                      <path
                                        d="M21 18v1c0 1.1-.9 2-2 2H5c-1.11 0-2-.9-2-2V5c0-1.1.89-2 2-2h14c1.1 0 2 .9 2 2v1h-9c-1.11 0-2 .9-2 2v8c0 1.1.89 2 2 2h9zm-9-2h10V8H12v8zm4-2.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
                                      />
                                    </svg>
                                    <span
                                      class="button__text"
                                    >
                                      420829
                                    </span>
                                  </button>
                                </form>
                              </section>
                            </div>
                          </article>
                        </div>
                      </div>
                    </section>
                  </div>
                </article>
              </section>
            </article>
          </div>
        </div>
      </section>
    </div>
  </article>
</DocumentFragment>
`;
